rust-runtime/aws-smithy-observability/src/meter.rs

```diff
@@ -0,0 +1,118 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+//! Metrics are used to gain insight into the operational performance and health of a system in
+//! real time.
+
+use crate::attributes::{Attributes, Context};
+use crate::error::ObservabilityError;
+
+/// Provides named instances of [Meter].
+pub trait MeterProvider {
+    /// Get or create a named [Meter].
+    fn get_meter(&self, scope: &'static str, attributes: Option<&Attributes>) -> Box<dyn Meter>;
+
+    /// Optional method to flush the metrics pipeline, default is noop
+    fn flush(&self) -> Result<(), ObservabilityError> {
```

Comment:
**correctness** Flush and shutdown aren't part of SRA, what purpose do these serve in practice? Why do we need them?

Also they don't feel like they belong here anyway as these are concerns of the consumer usually not of the library doing the instrumenting (which is the primary purpose/use of this API). In other words the instrumented code should not be responsible for flush/shutdown call.s

Reply:
I initially left these out of the trait and `impl`ed them on `AwsSdkOtelMeterProvider`. But at the end of your run when you do want to shutdown/flush the meter provider you first have to get it out of the `GlobalTelemetryProvider` and the only type information you get is `&dyn MetricProvider`. So to clear it you first have cast it to `Any`, then `downcast_ref` to the concrete type. All of that feels quite messy and not user friendly so I moved them to the actual trait. 

It seems likely that most metric solutions will provide some way to clear their buffers/shutdown gracefully (see the [clear method on buckets](https://docs.rs/metrics-util/0.5.0/metrics_util/struct.AtomicBucket.html#method.clear) in metrics.rs). 

So the addition of these turns this:
```rs
let otel_tp = set_global_telemetry_provider(None);
let foo = otel_tp
      .meter_provider()
      .as_any()
      .downcast_ref::<AwsSdkOtelMeterProvider>()
      .unwrap();
foo.shutdown().unwrap();
```

into:
```rs
let otel_tp = set_global_telemetry_provider(None);
otel_tp.meter_provider().shutdown().unwrap();
```

I'm not strongly opposed to removing them, but I opted for the more user friendly API to start.

Reply:
Removed in https://github.com/smithy-lang/smithy-rs/pull/3912/commits/0559e8b54fb41bcf04125e2e3cedeec87b24a9f9